File: ios/Demo/MekuriDemo/DemoScenery.swift
import SwiftUI

/// A row of buildings along the bottom of its area, with lit windows.
struct DemoSkyline: View {
    var body: some View {
        Canvas { context, size in
            var x: CGFloat = 0
            var seed = 7
            while x < size.width {
                seed = (seed * 1103515245 + 12345) % 2147483648
                let width = size.width * (0.06 + CGFloat(seed % 9) / 100)
                let height = size.height * (0.25 + CGFloat((seed / 9) % 45) / 100)
                let building = CGRect(x: x, y: size.height - height, width: width, height: height)
                context.fill(Path(building), with: .color(DemoInk.ink))
                let pane = max(3, width * 0.16)
                var wy = building.minY + pane
                while wy + pane < building.maxY {
                    var wx = building.minX + pane
                    while wx + pane < building.maxX {
                        seed = (seed * 1103515245 + 12345) % 2147483648
                        if seed % 3 != 0 {
                            context.fill(Path(CGRect(x: wx, y: wy, width: pane * 0.7, height: pane * 0.9)), with: .color(DemoInk.paper))
                        }
                        wx += pane * 1.6
                    }
                    wy += pane * 1.8
                }
                x += width + max(2, width * 0.12)
            }
            let ground = CGRect(x: 0, y: size.height - 3, width: size.width, height: 3)
            context.fill(Path(ground), with: .color(DemoInk.ink))
        }
    }
}

/// A moon: an ink ring with a toned crescent.
struct DemoMoon: View {
    var body: some View {
        GeometryReader { proxy in
            let side = min(proxy.size.width, proxy.size.height)
            ZStack {
                Circle()
                    .fill(DemoInk.paper)
                DemoHalftone(spacing: max(6, side * 0.045), opacity: 0.45)
                    .clipShape(Circle())
                    .mask {
                        Circle()
                            .offset(x: -side * 0.28, y: -side * 0.1)
                            .blendMode(.destinationOut)
                            .background(Circle())
                            .compositingGroup()
                    }
                Circle()
                    .stroke(DemoInk.ink, lineWidth: DemoInk.panelStroke)
            }
            .frame(width: side, height: side)
            .frame(maxWidth: .infinity, maxHeight: .infinity)
        }
    }
}

/// A mountain range along the bottom of its area, filled with tone.
struct DemoMountains: View {
    var body: some View {
        Canvas { context, size in
            let peaks: [CGPoint] = [
                CGPoint(x: 0.00, y: 0.70), CGPoint(x: 0.08, y: 0.45), CGPoint(x: 0.16, y: 0.62),
                CGPoint(x: 0.27, y: 0.28), CGPoint(x: 0.36, y: 0.55), CGPoint(x: 0.44, y: 0.40),
                CGPoint(x: 0.53, y: 0.66), CGPoint(x: 0.62, y: 0.33), CGPoint(x: 0.71, y: 0.58),
                CGPoint(x: 0.80, y: 0.22), CGPoint(x: 0.90, y: 0.50), CGPoint(x: 1.00, y: 0.64),
            ]
            var range = Path()
            range.move(to: CGPoint(x: 0, y: size.height))
            for peak in peaks {
                range.addLine(to: CGPoint(x: peak.x * size.width, y: peak.y * size.height))
            }
            range.addLine(to: CGPoint(x: size.width, y: size.height))
            range.closeSubpath()
            context.fill(range, with: .color(DemoInk.ink.opacity(0.85)))
            context.stroke(range, with: .color(DemoInk.ink), lineWidth: DemoInk.panelStroke)
        }
    }
}

File: android/demo/src/main/kotlin/app/mekuri/demo/DemoScenery.kt
package app.mekuri.demo

import androidx.compose.foundation.Canvas
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.clipPath
import kotlin.math.max
import kotlin.math.min

/** A row of buildings along the bottom of its area, with lit windows. */
@Composable
fun DemoSkyline(modifier: Modifier = Modifier) {
    Canvas(modifier) {
        var x = 0f
        var seed = 7L
        while (x < size.width) {
            seed = (seed * 1103515245 + 12345) % 2147483648L
            val width = size.width * (0.06f + (seed % 9).toFloat() / 100f)
            val height = size.height * (0.25f + ((seed / 9) % 45).toFloat() / 100f)
            val top = size.height - height
            drawRect(DemoInk.Ink, topLeft = Offset(x, top), size = Size(width, height))
            val pane = max(3f, width * 0.16f)
            var windowY = top + pane
            while (windowY + pane < size.height) {
                var windowX = x + pane
                while (windowX + pane < x + width) {
                    seed = (seed * 1103515245 + 12345) % 2147483648L
                    if (seed % 3 != 0L) {
                        drawRect(
                            DemoInk.Paper,
                            topLeft = Offset(windowX, windowY),
                            size = Size(pane * 0.7f, pane * 0.9f),
                        )
                    }
                    windowX += pane * 1.6f
                }
                windowY += pane * 1.8f
            }
            x += width + max(2f, width * 0.12f)
        }
        drawRect(DemoInk.Ink, topLeft = Offset(0f, size.height - 3f), size = Size(size.width, 3f))
    }
}

/** A moon: an ink ring with a toned crescent. */
@Composable
fun DemoMoon(modifier: Modifier = Modifier) {
    Canvas(modifier) {
        val side = min(size.width, size.height)
        val center = Offset(size.width / 2, size.height / 2)
        val radius = side / 2
        drawCircle(DemoInk.Paper, radius = radius, center = center)
        clipPath(circlePath(center, radius)) {
            drawHalftone(max(6f, side * 0.045f), 0.45f)
        }
        drawCircle(
            color = DemoInk.Paper,
            radius = radius,
            center = Offset(center.x - side * 0.28f, center.y - side * 0.1f),
        )
        drawCircle(
            color = DemoInk.Ink,
            radius = radius,
            center = center,
            style = Stroke(width = 3f),
        )
    }
}

/** A mountain range along the bottom of its area, filled with tone. */
@Composable
fun DemoMountains(modifier: Modifier = Modifier) {
    Canvas(modifier) {
        val peaks = listOf(
            0.00f to 0.70f, 0.08f to 0.45f, 0.16f to 0.62f,
            0.27f to 0.28f, 0.36f to 0.55f, 0.44f to 0.40f,
            0.53f to 0.66f, 0.62f to 0.33f, 0.71f to 0.58f,
            0.80f to 0.22f, 0.90f to 0.50f, 1.00f to 0.64f,
        )
        val range = Path().apply {
            moveTo(0f, size.height)
            peaks.forEach { (x, y) -> lineTo(x * size.width, y * size.height) }
            lineTo(size.width, size.height)
            close()
        }
        drawPath(range, DemoInk.Ink.copy(alpha = 0.85f))
        drawPath(range, DemoInk.Ink, style = Stroke(width = 3f))
    }
}

private fun circlePath(center: Offset, radius: Float) = Path().apply {
    addOval(
        Rect(
            left = center.x - radius,
            top = center.y - radius,
            right = center.x + radius,
            bottom = center.y + radius,
        ),
    )
}
